extension/jdtls.ext/com.microsoft.gradle.bs.importer/src/com/microsoft/gradle/bs/importer/GradleBuildServerBuildSupport.java

```diff
@@ -0,0 +1,238 @@
+                    new OutputPathsParams(Arrays.asList(buildTarget.getId()))).join();
+            String sourceOutputUri = getOutputUriByKind(outputResult.getItems(), OUTPUT_KIND_SOURCE);
+            IPath outputFullPath = getOutputFullPath(sourceOutputUri, project);
+            SourcesResult sourcesResult = buildServer.buildTargetSources(
+                    new SourcesParams(Arrays.asList(buildTarget.getId()))).join();
+            List<IClasspathEntry> sourceEntries = getSourceEntries(rootPath, project, sourcesResult, outputFullPath, isTest, monitor);
+            classpath.addAll(sourceEntries);
+
+            // TODO: resolve other classpath entries.
+        }
+
+        if (classpath.stream().anyMatch(cp -> cp.getEntryKind() == IClasspathEntry.CPE_SOURCE)) {
+            // if there is any source entry, we treat it as a Java project.
+            addJavaNature(project, monitor);
+            JavaCore.create(project).setRawClasspath(classpath.toArray(new IClasspathEntry[0]), monitor);
+            // refresh to let JDT be aware of the output folders.
+            // TODO: check if we only only refresh the output folder.
+            project.refreshLocal(IResource.DEPTH_INFINITE, monitor);
+        }
+    }
+
+
+
+    @Override
+    public boolean isBuildFile(IResource resource) {
+        return resource != null && resource.getType() == IResource.FILE && isBuildLikeFileName(resource.getName())
+                && Utils.isGradleBuildServerProject(resource.getProject());
+    }
+
+    @Override
+    public boolean isBuildLikeFileName(String fileName) {
+        return GRADLE_FILE_EXT.matcher(fileName).matches() || fileName.equals(GRADLE_PROPERTIES);
+    }
+
+    /**
+     * Return a list of classpath entries for all the source roots.
+     *
+     * @param rootPath the workspace root path.
+     * @param project the project.
+     * @param sourcesResult the sources result from build server.
+     * @param outputFullPath the output full path.
+     * @param isTest whether the source is test source.
+     * @param monitor the progress monitor.
+     * @throws CoreException
+     */
+    private List<IClasspathEntry> getSourceEntries(IPath rootPath, IProject project, SourcesResult sourcesResult, IPath outputFullPath, boolean isTest, IProgressMonitor monitor) throws CoreException {
+        List<IClasspathEntry> sourceEntries = new LinkedList<>();
+        for (SourcesItem sources : sourcesResult.getItems()) {
+            for (SourceItem source : sources.getSources()) {
+                IPath sourcePath = ResourceUtils.filePathFromURI(source.getUri());
+                IPath projectLocation = project.getLocation();
+                IPath sourceFullPath;
+                if (projectLocation.isPrefixOf(sourcePath)) {
+                    IPath relativeSourcePath = sourcePath.makeRelativeTo(project.getLocation());
+                    sourceFullPath = project.getFolder(relativeSourcePath).getFullPath();
+                } else {
```

Comment:
Does gradle support the case where the source folder is not under the project location?

Reply:
Yes, it's not common, but doable. See: https://stackoverflow.com/a/62693361/13068918

Reply:
If this is true, then it might contain multiple such kinds of source folders. You cannot link it to a same folder `_`.

Reply:
Updated to use String.join()